[CNSMR-3228] Handling deep links (#314)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/DeepLinks.md

- `babylon://homepage`
- `babylon://somescreen/parameter`
- `babylon://nested/screen/parameter`

Currently only single, simple parameters are supported app-wide as there was no requirement for anything else, and for reliability.

The prefix is not only `babylon://`, but also several others, and the app does not distinguish between them, looking only at the intent, as long as the OS is able to invoke the app. Thus, `babylon://link` and `otherprefix://link` are equivalent from the deep link handling standpoint within the app.

Let's trace the whole process from defining a new deep link intent, possibly restricting it to specific app targets, to presenting a new screen.

## Overall architecture

The central place to handle deep links is `ApplicationInvocationRouter`, instantiated during the app bootstrap, which takes a signal of `ApplicationInvocation` wrapping deep links and `NSUserActivity`.

Deep links are represented by the struct `DeepLink`:

``` swift
struct DeepLink: Decodable {
    let url: URL
    let options: [UIApplication.OpenURLOptionsKey : Any]

>     let intent: ApplicationInvocation.Intent?

Author: Oops, forgot the closing brace :)

`BabylonTabBarViewModel` in its turn produces a `Route.external(RoutingEvent, .deeplink)` event handled through its Flow controller, always presented modally over the tab bar controller. Some deep links switch tabs instead of presenting screens modally, which is encapsulated in the `RoutingEvent` as well.

Overall, the initial `DeepLink` value gets transformed into a `RoutingEvent` which is, essentially, just like a regular `Route` handled by Flow controllers in other parts of the app, and in `BabylonTabBarFlowController` these events are also handled in a regular fashion by constructing a screen with a `Builder` and presenting it.

### Prevent certain deep links from being handled

As a measure to release support for new deep links when the feature is fully complete, which is not necessarily at the time when the deep link is introduced, we have a special protocol:

``` swift
protocol DeeplinkConfiguration {
    func isEnabled(for intent: ApplicationInvocation.Intent) -> Bool
}
```

> Any app target can define a simple `var deeplinkConfiguration: DeeplinkConfiguration { get }` as part of its `AppConfiguration` and specify which deep links are not supported by that particular app flavour. These will be filtered out during handling in `ApplicationInvocationRouter` and will have no effect.

Reviewer: > as part of its `AppConfiguration`

🔥

The default value is just to return `true`, this enabling all available deep links.

### Open a deep link from within the app

> When the deep links found extended usage within the app, it became apparent that interacting with them in the app through `UIApplication.openURL(_:)` was extremely inconvenient during testing. It was common for both testers and developers to have multiple versions of the app installed at the same time, and iOS, as it turned out, randomly chooses the app to handle a particular custom URL scheme from the ones installed, as long as they support it. We were unable to control which app would be opened, when interacting with a deep link from inside an app, and this could even affect end users, since they could, in theory, have several of our apps installed at the same time.

Author: ```suggestion
When we started to use deep links in the app more, it became apparent that interacting with them in the app through `UIApplication.openURL(_:)` was extremely inconvenient during testing. It was common for both testers and developers to have multiple versions of the app installed, and iOS, as it turned out, randomly chooses the app to handle a particular custom URL scheme from the ones installed (as long as they support it). When interacting with a deep link from inside an app we were unable to control which app would be opened, and this could even affect end users, since in theory they could have several of our apps installed.
```

We created a special router to handle this situation:

``` swift
protocol TabDeeplinkRouterProtocol {
    func openDeeplink(_ deeplink: DeepLink)
    func handle(intent: ApplicationInvocation.Intent)
}
```

Internally it simply forwards the deep link interaction to the instance of `ApplicationInvocationRouter`, removing the OS from the equation and making sure that the deep link is always opened in the same app that it was interacted in.

If you're creating a new screen or just want to open an existing screen by using a deep link without the need to construct it, use `TabDeeplinkRouterProtocol` instead of `UIApplication.openURL(_:)` to avoid the problem described above.

## Create a new deep link

Let's dive into how to define, use and test a new deep link.

### Intent and routing event

First off, we need to introduce a new intent, which is the part of the deeplink after the prefix:

``` swift
public enum ApplicationInvocation {
    /// Custom Scheme URL.
    case deeplink(DeepLink)

>     /// - note: `rawValue` is URL host name.

Author: It's in the code, so might as well keep it.
